File: src/main/java/com/qtech/bubbles/common/AttributeMap.java
package com.qtech.bubbles.common;

import com.qtech.bubbles.common.entity.Attribute;
import com.qtech.bubbles.common.function.BiFloat2FloatFunction;
import com.qtech.bubbles.common.function.Float2FloatFunction;
import com.qtech.bubbles.common.holders.IArrayDataHolder;
import org.bson.*;

import javax.annotation.Nullable;
import java.util.*;

public class AttributeMap implements IArrayDataHolder<AttributeMap> {
    private final HashMap<Attribute, Float> map = new HashMap<>();

    public AttributeMap() {

    }

    public void setBase(Attribute attribute, float value) {
        this.map.put(attribute, value);
    }

    public float getBase(Attribute attribute) {
        @Nullable Float value = this.map.get(attribute);
        if (value == null) {
            throw new NoSuchElementException("Attribute \"" + attribute.getName() + "\" has no set base value.");
        }
        return value;
    }

    public void getModified(Attribute attribute, BiFloat2FloatFunction function, List<AttributeMap> attributeMaps) {
        if (!this.map.containsKey(attribute)) {
            throw new NoSuchElementException("Attribute \"" + attribute.getName() + "\" has no set base value.");
        }

        float f = getBase(attribute);
        for (AttributeMap map : attributeMaps) {
            f *= function.apply(f, map.getBase(attribute));
        }
    }

    public void getModified(Attribute attribute, BiFloat2FloatFunction function, AttributeMap... attributeMaps) {
        if (!this.map.containsKey(attribute)) {
            throw new NoSuchElementException("Attribute \"" + attribute.getName() + "\" has no set base value.");
        }

        float f = getBase(attribute);
        for (AttributeMap map : attributeMaps) {
            f *= function.apply(f, map.getBase(attribute));
        }
    }

    public void getModified(Attribute attribute, List<Float2FloatFunction> functions) {
        if (!this.map.containsKey(attribute)) {
            throw new NoSuchElementException("Attribute \"" + attribute.getName() + "\" has no set base value.");
        }

        float f = getBase(attribute);
        for (Float2FloatFunction function : functions) {
            f *= function.apply(f);
        }
    }

    public void getModified(Attribute attribute, Float2FloatFunction... functions) {
        if (!this.map.containsKey(attribute)) {
            throw new NoSuchElementException("Attribute \"" + attribute.getName() + "\" has no set base value.");
        }

        float f = getBase(attribute);
        for (Float2FloatFunction function : functions) {
            f *= function.apply(f);
        }
    }

    public BsonArray write(BsonArray array) {
        for (Map.Entry<Attribute, Float> entry : this.map.entrySet()) {
            BsonDocument document = new BsonDocument();
            document.put("name", new BsonString(entry.getKey().getName()));
            document.put("value", new BsonDouble(entry.getValue()));
            array.add(document);
        }
        return array;
    }

    public AttributeMap read(BsonArray array) {
        for (BsonValue item : array.getValues()) {
            if (item instanceof BsonDocument) {
                BsonDocument document = (BsonDocument) item;
                Attribute key = Attribute.fromName(document.getString("name").getValue());
                float value = (float) document.getDouble("value").getValue();

                this.map.put(key, value);
            }
        }
        return this;
    }

    public void setAll(AttributeMap defaultAttributes) {
        this.map.putAll(defaultAttributes.map);
    }
}

File: src/main/kotlin/qtech/bubbles/common/AttributeMap.kt
package qtech.bubbles.common

import qtech.bubbles.common.entity.Attribute
import qtech.bubbles.common.holders.IArrayDataHolder
import net.querz.nbt.tag.CompoundTag
import net.querz.nbt.tag.ListTag

class AttributeMap : IArrayDataHolder<CompoundTag?, AttributeMap?> {
    private val map = HashMap<Attribute?, Float>()
    operator fun set(attribute: Attribute?, value: Float) {
        map[attribute] = value
    }

    operator fun get(attribute: Attribute): Float {
        return map[attribute] ?: throw NoSuchElementException("Attribute \"" + attribute.name + "\" has no set value.")
    }

    fun getModified(attribute: Attribute, function: (Float, Float) -> Float, attributeMaps: List<AttributeMap>) {
        if (!map.containsKey(attribute)) {
            throw NoSuchElementException("Attribute \"" + attribute.name + "\" has no set base value.")
        }
        var f = get(attribute)
        for (map in attributeMaps) {
            f *= function.invoke(f, map.get(attribute))
        }
    }

    fun getModified(attribute: Attribute, function: (Float, Float) -> Float, vararg attributeMaps: AttributeMap) {
        if (!map.containsKey(attribute)) {
            throw NoSuchElementException("Attribute \"" + attribute.name + "\" has no set base value.")
        }
        var f = get(attribute)
        for (map in attributeMaps) {
            f *= function.invoke(f, map.get(attribute))
        }
    }

    fun getModified(attribute: Attribute, functions: List<(Float) -> Float>) {
        if (!map.containsKey(attribute)) {
            throw NoSuchElementException("Attribute \"" + attribute.name + "\" has no set base value.")
        }
        var f = get(attribute)
        for (function in functions) {
            f *= function.invoke(f)
        }
    }

    fun getModified(attribute: Attribute, vararg functions: (Float) -> Float) {
        if (!map.containsKey(attribute)) {
            throw NoSuchElementException("Attribute \"" + attribute.name + "\" has no set base value.")
        }
        var f = get(attribute)
        for (function in functions) {
            f *= function.invoke(f)
        }
    }

    override fun write(listTag: ListTag<CompoundTag?>): ListTag<CompoundTag?> {
        for ((key, value) in map) {
            val document = CompoundTag()
            document.putString("name", key!!.name)
            document.putFloat("value", value)
            listTag.add(document)
        }
        return listTag
    }

    override fun read(listTag: ListTag<CompoundTag?>): AttributeMap {
        for (item in listTag) {
            val key = Attribute.fromName(item!!.getString("name"))
            val value = item.getDouble("value").toFloat()
            map[key] = value
        }
        return this
    }

    fun setAll(defaultAttributes: AttributeMap) {
        map.putAll(defaultAttributes.map)
    }
}